File: tests/test_phase19_fleet.js
/**
 * Phase 19.2 — Operational Fleet Monitoring & Evidence Integration Test Suite (Node.js)
 * =====================================================================================
 */

'use strict';

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const assert = require('assert');

const ROOT = path.resolve(__dirname, '..');
const { FleetManagerJS, VALID_EQUIPMENT_IDS } = require('../src/fleet/fleet_manager');

function computeSha256(filePath) {
  const content = fs.readFileSync(filePath, 'utf8').replace(/\r\n/g, '\n');
  return crypto.createHash('sha256').update(content, 'utf8').digest('hex');
}

console.log('--- PREDICTA-26 Phase 19.2 Fleet Monitoring Test Suite (Node.js) ---');

// 1. Protected Model & Dataset Integrity
const modelSha = computeSha256(path.join(ROOT, 'ml/models/production/predicta_xgboost_model.json'));
assert.strictEqual(modelSha, '91bb598ae91155674e40cb0a9f39d1e9bdeacd39875542db88b65e3668f29d98', 'Model SHA mismatch');
console.log('✓ Protected Model SHA-256 Verified');

const datasetSha = computeSha256(path.join(ROOT, 'ml/data/synthetic/predicta_dataset_v3_50000.csv'));
assert.strictEqual(datasetSha, '48e718643b6fe99bc410421f5b48715c294f1c4c2edabf10870935afdb820a06', 'Dataset SHA mismatch');
console.log('✓ Protected Dataset SHA-256 Verified');

// 2. Frontend Mirror Byte Parity
const pairs = [
  ['index.html', 'frontend/index.html'],
  ['script.js', 'frontend/script.js'],
  ['api.js', 'frontend/api.js']
];
for (const [r, f] of pairs) {
  const rBuf = fs.readFileSync(path.join(ROOT, r));
  const fBuf = fs.readFileSync(path.join(ROOT, f));
  assert.strictEqual(rBuf.length, fBuf.length, `Byte length mismatch on ${r} vs ${f}`);
  assert.strictEqual(
    crypto.createHash('sha256').update(rBuf).digest('hex'),
    crypto.createHash('sha256').update(fBuf).digest('hex'),
    `Hash mismatch on ${r} vs ${f}`
  );
}
console.log('✓ Frontend Exact Mirror Byte Parity Verified (3/3 pairs)');

// 3. Zero SHAP & Prohibited Multipliers
const feFiles = ['index.html', 'frontend/index.html', 'script.js', 'frontend/script.js', 'api.js', 'frontend/api.js'];
const shapRegex = /\bshap\b/i;
const multRegexes = [/\*\s*1\.2\b/, /\*\s*1\.05\b/, /\*\s*0\.95\b/];

for (const rel of feFiles) {
  const content = fs.readFileSync(path.join(ROOT, rel), 'utf8');
  assert(!shapRegex.test(content), `Prohibited SHAP token in ${rel}`);
  for (const rx of multRegexes) {
    assert(!rx.test(content), `Prohibited client multiplier in ${rel}`);
  }
}
console.log('✓ Zero SHAP and Multipliers in Frontend Code Verified');

// 4. Fleet Monitoring Dashboard DOM Contract
const html = fs.readFileSync(path.join(ROOT, 'index.html'), 'utf8');
const requiredIds = [
  'fleet-monitoring-dashboard',
  'fleet-total-lots',
  'fleet-total-wafers',
  'fleet-total-components',
  'fleet-total-equipment',
  'fleet-operating-threshold',
  'fleet-cohort-filter',
  'btn-fleet-refresh',
  'fleet-lots-table',
  'fleet-lots-tbody',
  'fleet-lot-detail-panel',
  'fleet-selected-lot-id',
  'fleet-selected-lot-cohort',
  'fleet-selected-lot-station',
  'fleet-selected-lot-wafers',
  'fleet-selected-lot-components',
  'btn-fleet-close-detail'
];
for (const id of requiredIds) {
  assert(html.includes(`id="${id}"`), `Missing required DOM ID: ${id}`);
}
console.log('✓ Fleet Monitoring Dashboard DOM Contract Verified');

// 5. Fleet Manager Summary & Lot Queries
const mgr = new FleetManagerJS();
const summary = mgr.getFleetSummary();
assert.strictEqual(summary.total_lots, 50, 'Total lots must be 50');
assert(summary.total_wafers >= 100, 'Total wafers must be >= 100');
assert.strictEqual(summary.total_components, 5000, 'Total components must be 5000');
assert.strictEqual(summary.total_equipment, 5, 'Total equipment stations must be 5');
assert.strictEqual(summary.provenance.operating_threshold, 0.20, 'Operating threshold must be 0.20');
assert.strictEqual(summary.lot_cohort_distribution.TRAIN, 35, 'Train cohort must have 35 lots');
assert.strictEqual(summary.lot_cohort_distribution.TEST, 8, 'Test cohort must have 8 lots');
console.log('✓ Fleet Manager Summary Resolution Verified');

// 6. Lot and Wafer Details
const lot1 = mgr.getLotDetail('LOT-SYN-001');
assert(lot1 !== null);
assert.strictEqual(lot1.cohort_type, 'TRAIN');
assert(lot1.canonical_components.length === 3);
assert(lot1.wafers.includes('W-2026-01'));
console.log('✓ Demo Lot LOT-SYN-001 Canonical Links Verified');

const wfr1 = mgr.getWaferDetail('WFR-001');
assert(wfr1 !== null);
assert.strictEqual(wfr1.lot_id, 'LOT-SYN-001');
assert.strictEqual(wfr1.die_count, 50);
console.log('✓ Wafer Detail Query Resolution Verified');

// 7. Fail-closed behavior
assert.strictEqual(mgr.getLotDetail('INVALID_LOT'), null);
assert.strictEqual(mgr.getLotDetail(''), null);
assert.strictEqual(mgr.getLotDetail(null), null);
assert.strictEqual(mgr.getWaferDetail('INVALID_WFR'), null);
assert.strictEqual(mgr.getWaferDetail(''), null);
assert.strictEqual(mgr.getWaferDetail(null), null);
console.log('✓ Fail-Closed Behavior on Unregistered Entities Verified');

console.log('\n========================================');
console.log('ALL PHASE 19.2 NODE.JS TESTS PASSED (7/7)');
console.log('========================================\n');

File: tests/test_phase19_fleet.py
"""
Phase 19.2 — Operational Fleet Monitoring & Evidence Integration Test Suite (Python)
=====================================================================================
Validates:
1. Protected Artifacts (Model SHA-256, Dataset SHA-256, Threshold 0.20)
2. Frontend Byte Parity (index.html, script.js, api.js mirrors)
3. Zero SHAP & Prohibited Multipliers in Frontend Code
4. Fleet Monitoring Dashboard DOM Contract
5. Authoritative Fleet Manager Engine (50 Lots, 100 Wafers, 5000 Dies, 5 Stations)
6. Disjoint Cohort Splits & Canonical Genealogy Resolution
7. Fail-Closed Behavior for Invalid / Unregistered Entities
"""

import hashlib
import json
import re
from pathlib import Path

from src.fleet.fleet_manager import FleetManagerPy, VALID_EQUIPMENT_IDS

ROOT = Path(__file__).resolve().parent.parent

PROD_MODEL_PATH = ROOT / "ml" / "models" / "production" / "predicta_xgboost_model.json"
PROD_DATASET_PATH = ROOT / "ml" / "data" / "synthetic" / "predicta_dataset_v3_50000.csv"

PROTECTED_MODEL_SHA = "91bb598ae91155674e40cb0a9f39d1e9bdeacd39875542db88b65e3668f29d98"
PROTECTED_DATASET_SHA = "48e718643b6fe99bc410421f5b48715c294f1c4c2edabf10870935afdb820a06"
PROTECTED_THRESHOLD = 0.20


def compute_sha256(path: Path) -> str:
    content = path.read_text(encoding="utf-8").replace("\r\n", "\n")
    return hashlib.sha256(content.encode("utf-8")).hexdigest()


def test_protected_model_sha256():
    assert PROD_MODEL_PATH.exists(), f"Model file missing at {PROD_MODEL_PATH}"
    computed_sha = compute_sha256(PROD_MODEL_PATH)
    assert computed_sha == PROTECTED_MODEL_SHA, (
        f"Model SHA mismatch: expected {PROTECTED_MODEL_SHA}, got {computed_sha}"
    )


def test_protected_dataset_sha256():
    assert PROD_DATASET_PATH.exists(), f"Dataset file missing at {PROD_DATASET_PATH}"
    computed_sha = compute_sha256(PROD_DATASET_PATH)
    assert computed_sha == PROTECTED_DATASET_SHA, (
        f"Dataset SHA mismatch: expected {PROTECTED_DATASET_SHA}, got {computed_sha}"
    )


def test_protected_operating_threshold():
    contract_path = ROOT / "ml" / "reliability_twin" / "reliability_twin_contract.json"
    assert contract_path.exists()
    contract = json.loads(contract_path.read_text(encoding="utf-8"))
    assert contract["immutability_constraints"]["authoritative_operating_threshold"] == PROTECTED_THRESHOLD


def test_frontend_exact_byte_parity():
    pairs = [
        ("index.html", "frontend/index.html"),
        ("script.js", "frontend/script.js"),
        ("api.js", "frontend/api.js"),
    ]
    for root_rel, fe_rel in pairs:
        root_file = ROOT / root_rel
        fe_file = ROOT / fe_rel
        assert root_file.exists(), f"Root file {root_file} missing"
        assert fe_file.exists(), f"Frontend mirror {fe_file} missing"
        root_bytes = root_file.read_bytes()
        fe_bytes = fe_file.read_bytes()
        assert len(root_bytes) == len(fe_bytes), (
            f"Size mismatch {root_rel} ({len(root_bytes)}) vs {fe_rel} ({len(fe_bytes)})"
        )
        assert hashlib.sha256(root_bytes).hexdigest() == hashlib.sha256(fe_bytes).hexdigest(), (
            f"Hash mismatch between {root_rel} and {fe_rel}"
        )


def test_frontend_zero_shap_and_multipliers():
    frontend_files = [
        ROOT / "index.html",
        ROOT / "frontend" / "index.html",
        ROOT / "script.js",
        ROOT / "frontend" / "script.js",
        ROOT / "api.js",
        ROOT / "frontend" / "api.js",
    ]
    shap_pattern = re.compile(r"\bshap\b", re.IGNORECASE)
    prohibited_multipliers = [r"\*\s*1\.2\b", r"\*\s*1\.05\b", r"\*\s*0\.95\b"]

    for file_path in frontend_files:
        content = file_path.read_text(encoding="utf-8")
        assert not shap_pattern.search(content), (
            f"Prohibited SHAP token found in {file_path.name}"
        )
        for pattern in prohibited_multipliers:
            assert not re.search(pattern, content), (
                f"Prohibited client multiplier '{pattern}' found in {file_path.name}"
            )


def test_fleet_monitoring_dashboard_dom_contract():
    html_content = (ROOT / "index.html").read_text(encoding="utf-8")

    # Container
    assert 'id="fleet-monitoring-dashboard"' in html_content

    # KPI elements
    required_kpi_ids = [
        "fleet-total-lots",
        "fleet-total-wafers",
        "fleet-total-components",
        "fleet-total-equipment",
        "fleet-operating-threshold",
    ]
    for dom_id in required_kpi_ids:
        assert f'id="{dom_id}"' in html_content, f"Missing KPI DOM element: {dom_id}"

    # Controls & Table
    required_control_ids = [
        "fleet-cohort-filter",
        "btn-fleet-refresh",
        "fleet-lots-table",
        "fleet-lots-tbody",
        "fleet-lot-detail-panel",
        "fleet-selected-lot-id",
        "fleet-selected-lot-cohort",
        "fleet-selected-lot-station",
        "fleet-selected-lot-wafers",
        "fleet-selected-lot-components",
        "btn-fleet-close-detail",
    ]
    for dom_id in required_control_ids:
        assert f'id="{dom_id}"' in html_content, f"Missing control DOM element: {dom_id}"


def test_fleet_manager_summary_resolution():
    mgr = FleetManagerPy()
    summary = mgr.get_fleet_summary()

    assert summary["fleet_id"] == "PREDICTA_FLEET_QUALIFICATION_COHORT_2026"
    assert summary["total_lots"] == 50
    assert summary["total_wafers"] >= 100
    assert summary["total_components"] == 5000
    assert summary["total_equipment"] == 5
    assert summary["provenance"]["operating_threshold"] == 0.20
    assert summary["provenance"]["is_synthetic"] is True

    # Check cohort breakdown
    cohort_dist = summary["lot_cohort_distribution"]
    assert cohort_dist["TRAIN"] == 35
    assert cohort_dist["VALIDATION_TUNE"] == 3
    assert cohort_dist["CALIBRATION"] == 4
    assert cohort_dist["TEST"] == 8


def test_fleet_manager_lot_queries():
    mgr = FleetManagerPy()
    lots = mgr.get_fleet_lots()
    assert len(lots) == 50

    # Test LOT-SYN-001 (Demo lot with canonical components)
    lot1 = mgr.get_lot_detail("LOT-SYN-001")
    assert lot1 is not None
    assert lot1["lot_id"] == "LOT-SYN-001"
    assert lot1["cohort_type"] == "TRAIN"
    assert lot1["equipment_id"] in VALID_EQUIPMENT_IDS
    assert "WFR-001" in lot1["wafers"]
    assert "WFR-002" in lot1["wafers"]
    assert "W-2026-01" in lot1["wafers"]
    assert len(lot1["canonical_components"]) == 3
    comp_ids = [c["component_id"] for c in lot1["canonical_components"]]
    assert "COMP-NORMAL" in comp_ids
    assert "COMP-LATENT_DEFECT" in comp_ids
    assert "COMP-FALSE_ALARM" in comp_ids

    # Test Test cohort lot
    lot45 = mgr.get_lot_detail("LOT-SYN-045")
    assert lot45 is not None
    assert lot45["cohort_type"] == "TEST"
    assert lot45["component_count"] == 100


def test_fleet_manager_wafer_queries():
    mgr = FleetManagerPy()

    wfr1 = mgr.get_wafer_detail("WFR-001")
    assert wfr1 is not None
    assert wfr1["wafer_id"] == "WFR-001"
    assert wfr1["lot_id"] == "LOT-SYN-001"
    assert wfr1["die_count"] == 50

    wfrDemo = mgr.get_wafer_detail("W-2026-01")
    assert wfrDemo is not None
    assert wfrDemo["lot_id"] == "LOT-SYN-001"
    assert "COMP-NORMAL" in wfrDemo["canonical_components"]


def test_fleet_manager_fails_closed_on_invalid_entities():
    mgr = FleetManagerPy()

    assert mgr.get_lot_detail("LOT-INVALID-999") is None
    assert mgr.get_lot_detail("") is None
    assert mgr.get_lot_detail(None) is None

    assert mgr.get_wafer_detail("WFR-INVALID-999") is None
    assert mgr.get_wafer_detail("") is None
    assert mgr.get_wafer_detail(None) is None


def test_script_js_exports_render_fleet_monitoring():
    script_content = (ROOT / "script.js").read_text(encoding="utf-8")
    assert "window.renderFleetMonitoringDashboard = renderFleetMonitoringDashboard;" in script_content
